File: shellSort.c
/* Sooraj Modi
 * CIS 3190 - Assignment 4
 * modis@uoguelph.ca
 * 0965941
 * To compile: "gcc --Wall -std=c99 shellSort.c -o output"
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

//void generateRandomArray(int** numArr);
void readArrayFromFile(int** numArr, int* arrLen, FILE* fp);
void shellSort(int** numArr, int* arrLen);
void outputResults(int** numArr, int* arrLen, char* fileName);

int main(void){
  printf("Beginning the Shell Sort Program in C\n");

  int* numArr = NULL;

  char fileName[50];
  int* arrLen = NULL;
  arrLen = malloc(sizeof(int));
  FILE* fp = NULL;

  // Open file to be read
  printf("Enter the name of the file of integers to be sorted: ");
  fgets(fileName, 50, stdin);
  fileName[strlen(fileName) - 1] = '\0';
  fp = fopen(fileName, "r");
  if(!fp){
    printf("Unable to find file\nNow terminating program\n");
    return 0;
  }

  readArrayFromFile(&numArr, arrLen, fp);

  //Commence algorithm and timer
  time_t start, stop;
  start=clock();
  printf("Starting Timer\nBeginning shell sort\n");
  shellSort(&numArr, arrLen);
  printf("Successfully sorted array\nStopping Timer\n");
  stop=clock();
  printf("Total Time for ShellSort subroutine is: %ld micro seconds\n", (long int)(stop - start));

  //Output results
  printf("Outputting results to sorted.txt\n");
  outputResults(&numArr, arrLen, "sortedC.txt");

  //Free allocated memory
  if(!numArr) free(numArr);
  if(!arrLen) free(arrLen);
  fclose(fp);

  printf("Successfully exiting program\n");
  return 0;
}

void shellSort(int** numArr, int* arrLen){
  // Beginning shell sort algorithm
  for(int gap = *arrLen/2; gap > 0; gap /= 2){
    for(int i = gap; i < *arrLen; i++){
      int temp = (*numArr)[i];
      int j = i;

      for(j = i; j >= gap && (*numArr)[j - gap] > temp; j--){
        (*numArr)[j] = (*numArr)[j-gap];
      }

      (*numArr)[j] = temp;
    }
  }
}

void readArrayFromFile(int** numArr, int* arrLen, FILE* fp){
  if(!fp)  return;
  
  fseek(fp, 0, SEEK_END);
  long int eof = ftell(fp);
  fseek(fp, 0, SEEK_SET);

  *numArr = malloc(sizeof(int)*100500);

  char input[30];
  int index = 0;

  // Read every element in file and place into array
  while(ftell(fp) != eof){
    fgets(input, 30, fp);
    (*numArr)[index] = atoi(input);
    index++;
  }

  *arrLen = index;
}

void outputResults(int** numArr, int* arrLen, char* fileName){
	FILE* fp = NULL;
	fp = fopen(fileName, "w");
	//Check to make sure file was made (error proofing)
	if(!fp){
		printf("Unable to create a new file for writing\nWriting to file failed\n");
		return;
	}

    // Output every element of the array into a file
	for(int i = 0; i < *arrLen; i++)  fprintf(fp, "%d\n", (*numArr)[i]);

	fclose(fp);
}

File: shellSort.py
# Sooraj Modi
# 0965941
# modis@uoguelph.ca
# CIS 3190 - Assignment 4
# To compile: python shellSort.py

import time
import random
from array import *

def readListFromFile(numList):
	fileName = raw_input("Enter the name of the file to be read: ");
	# Open and read numbers from file into array
	fp = open(fileName, "r");
	for line in fp:
		line = line.replace('\n','')
		numList.append(line);	
	fp.close()
	print("File successfully read!");
	
def writeArrayToFile(numArr):
    print("Outputting sorted array to sortedPy.txt");
    # Open and output to file
    with open("sortedPy.txt", "w") as fp:
		for item in numArr:
			fp.write("%s\n" % item)
    fp.close()
    print("Successfully printed array to file");
    
def convertListToArray(numList, numArray):
    newList = ';'.join(numList)
    for item in newList.split(';'):
	    numArr.append(item)
    
def shellSort(numArr):
    print("Beginning shell sort algorithm")
    print("Beginning timer")
    begin = time.time()
    
    # Begin shell sort algorithm
    arrLen = len(numArr)
    gap = arrLen / 2
    while gap > 0:
        for i in range(gap, arrLen):
            temp = numArr[i]
            j = i
			
            while ((j >= gap) and (numArr[j-gap] > temp)):
                numArr[j] = numArr[j-gap]
                j = j - gap
			
            numArr[j] = temp
        gap //= 2
    
    # Output run time    
    print("Stopping timer")
    end = time.time()
    print("Total algorithm run time: ", end-begin, "seconds")    

# Say Hello
print("Beginning shell sort program in python")

# Read from file
numList = list();
readListFromFile(numList)

# Convert list to array
numArr = []
convertListToArray(numList, numArr)

# Sort the array
shellSort(numArr)

# Write to file
writeArrayToFile(numArr)

# Say goodbye <- pls give me full comment marks
print("Successfully exiting program")
